src/jit/codegen.h

```diff
@@ -1449,7 +1449,6 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
         static const GenConditionDesc& Get(GenCondition condition)
         {
-            assert(condition.GetCode() < _countof(map));
```

Comment:
This is good assert to have. What if `map` definition changes?

Reply:
Most of these were removed for clang 3.9 in fdf04aa. Without this commit (and with `tutological-compare` warning enabled), gcc5-8 and clang6-8 are satisfied. I can move `tautological-compare` changes in a separate PR and keep this one for nonnull-comparison (which mainly focuses on `this == null`)?

I think the warning was `Wtautological-constant-out-of-range-compare`, as it leads to dead code. 

> What if map definition changes?

I think that is the reason why they have stopped complaining about it in clang6. :)

Reply:
> I think that is the reason why they have stopped complaining about it in clang6

I would keep `tautological-compare` warning disabled then since it is not helping in its current state.

Reply:
@jkotas, @janvorli, I have disabled `tautological-compare` only where compiler is too sensitive (feature detection in `configure.cmake` and used on `configurecompiler.cmake`). The introspection check succeeds on clang v6+ (tested with v6 and latest stable v8, both succeeded) and fails on v3.9 to v5.x on Linux/FreeBSD and corresponding versions of Apple Clang (which has different versioning scheme). The reason to keep it for higher versions of clang is that it has provided some useful dead-code detection: f84aa60943b36d0840bc337e39dbd87ce7af38f8.

Reply:
It seems that the tautological-compare warning is actually a superset of several tautological warnings. Here is the hierarchy extracted from https://github.com/llvm-mirror/clang/blob/d8066f9602863a39f208a52ca3be977c46a5d0f0/include/clang/Basic/DiagnosticGroups.td:
* tautological-compare
  * tautological-pointer-compare
  * tautological-constant-compare
    * tautological-constant-out-of-range-compare
  * tautological-overlap-compare
  * tautological-undefined-compare

So maybe we can use the more fine-grained one(s) to disable what we really need.

Reply:
Unfortunately, the top-level warning is also thrown by Clang 3.9, as well as the deepest `tautological-constant-out-of-range-compare`:

```
/datadrive/projects/coreclr2/src/md/enc/stgio.cpp:645:47: error: comparison of unsigned expression < 0 is always false [-Werror,-Wtautological-compare]
                if (cbRtn > m_cbData || cbRtn < 0)
                                        ~~~~~ ^ ~
...
/datadrive/projects/coreclr2/src/jit/_typeinfo.h:334:23: error: result of comparison of constant 63 with expression of type 'ti_types' is always true [-Werror,-Wtautological-constant-out-of-range-compare]
        assert(tiType <= TI_FLAG_DATA_MASK);
        ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~
```

Reply:
I think it is fine to compile the code-base with higher warning levels from time to time and fix what makes sense. But I do not think we should bother with having these questionable warnings enabled by default. They just cause pain when they are enabled by default.

Reply:
@jkotas, since <kbd>[release/3.0](https://github.com/dotnet/coreclr/tree/release/3.0)</kbd> has been branched out, if master is getting post-3.0 fixes could this PR be defrosted?
I have some Solaris patches lining up which are done with gcc7 on top of these changes.

Reply:
We are still merging all fixes from master to release/3.0. release/3.0 is used to build the official preview builds while there is ongoing work on 3.0 in master.

Reply:
In the current shape of this change, there seems to be no room for fallout, compared to other changes been made to master. Do you still think it has to made or can this be merged?

Reply:
> there seems to be no room for fallout

I think it still has opportunity for fallout. It is easy to miss cases in changes like these.

> other changes been made to master

The changes going to master at this point should be either important for 3.0 or no risk. I am sorry, but this one is not important for 3.0.

Reply:
It's OK. I am unblocked with change made in master disabling this warning for gcc. This was anyway nice to have cleanup. Thanks for your review.